src/esp/gfx/ssao/HBAOHelper.cpp

```diff
@@ -0,0 +1,689 @@
+
+    nvgl::bindMultiTexture(GL_TEXTURE0, GL_TEXTURE_2D,
+                           textures.scene_depthlinear);
+    glDrawArrays(GL_TRIANGLES, 0, 3);
+    nvgl::bindMultiTexture(GL_TEXTURE0, GL_TEXTURE_2D, 0);
+  }
+
+  {
+    // NV_PROFILE_GL_SECTION("deinterleave");
+    glBindFramebuffer(GL_FRAMEBUFFER, fbos.hbao2_deinterleave);
+    glViewport(0, 0, quarterWidth, quarterHeight);
+
+    glUseProgram(m_progManager.get(programs.hbao2_deinterleave));
+    nvgl::bindMultiTexture(GL_TEXTURE0, GL_TEXTURE_2D,
+                           textures.scene_depthlinear);
+
+    for (int i = 0; i < HBAO_RANDOM_ELEMENTS; i += NUM_MRT) {
+      glUniform4f(0, float(i % 4) + 0.5f, float(i / 4) + 0.5f,
+                  m_hbaoUbo.InvFullResolution.x, m_hbaoUbo.InvFullResolution.y);
+
+      for (int layer = 0; layer < NUM_MRT; layer++) {
+        glFramebufferTexture(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0 + layer,
+                             textures.hbao2_depthview[i + layer], 0);
+      }
+      glDrawArrays(GL_TRIANGLES, 0, 3);
+    }
+  }
+
+  {
+    // NV_PROFILE_GL_SECTION("ssaocalc");
+
+    glBindFramebuffer(GL_FRAMEBUFFER, fbos.hbao2_calc);
+    glViewport(0, 0, quarterWidth, quarterHeight);
+
+    glUseProgram(m_progManager.get(USE_AO_SPECIALBLUR && m_config.blur
+                                       ? programs.hbao2_calc_blur
+                                       : programs.hbao2_calc));
+    nvgl::bindMultiTexture(GL_TEXTURE1, GL_TEXTURE_2D,
+                           textures.scene_viewnormal);
+
+    glBindBufferBase(GL_UNIFORM_BUFFER, 0, buffers.hbao_ubo);
+    glNamedBufferSubData(buffers.hbao_ubo, 0, sizeof(HBAOData), &m_hbaoUbo);
+
+#if USE_AO_LAYERED_SINGLEPASS
+    // instead of drawing to each layer individually
+    // we draw all layers at once, and use image writes to update the array
+    // texture this buys additional performance :)
+
+    nvgl::bindMultiTexture(GL_TEXTURE0, GL_TEXTURE_2D_ARRAY,
+                           textures.hbao2_deptharray);
+#if USE_AO_LAYERED_SINGLEPASS == AO_LAYERED_IMAGE
+    glBindImageTexture(0, textures.hbao2_resultarray, 0, GL_TRUE, 0,
+                       GL_WRITE_ONLY, USE_AO_SPECIALBLUR ? GL_RG16F : GL_R8);
+#endif
+    glDrawArrays(GL_TRIANGLES, 0, 3 * HBAO_RANDOM_ELEMENTS);
+#if USE_AO_LAYERED_SINGLEPASS == AO_LAYERED_IMAGE
+    glMemoryBarrier(GL_TEXTURE_FETCH_BARRIER_BIT |
+                    GL_SHADER_IMAGE_ACCESS_BARRIER_BIT);
+#endif
+#else
+    for (int i = 0; i < HBAO_RANDOM_ELEMENTS; i++) {
+      glUniform2f(0, float(i % 4) + 0.5f, float(i / 4) + 0.5f);
+      glUniform4fv(1, 1, m_hbaoRandom[i].get_value());
+
+      nvgl::bindMultiTexture(GL_TEXTURE0, GL_TEXTURE_2D,
+                             textures.hbao2_depthview[i]);
+      glFramebufferTextureLayer(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0,
+                                textures.hbao2_resultarray, 0, i);
+
+      glDrawArrays(GL_TRIANGLES, 0, 3);
+    }
+#endif
+  }
```

Comment:
This would be something like

```cpp
// Full-screen triangle(s). Construct them once and reuse.
Mn::GL::Mesh fullScreenTriangle;
fullScreenTriangle.setCount(3);
Mn::GL::Mesh fullScreenTriangles;
fullScreenTriangles.setCount(3*HBAO_RANDOM_ELEMENTS);

// Assuming the framebuffer exists already
Mn::GL::Framebuffer& hbao2CalcFramebuffer = ...;
hbao2CalcFramebuffer
  // setViewport() doesn't need to be called if a correct viewport was passed
  // to hbao2CalcFramebuffer constructor already
  .setViewport({{}, {quarterWidth, quarterHeight}})
  .bind();

hbaoControlBuffer.setSubData(0, m_hbaoUbo);

// Assuming the shader exists already
HbaoShader& shader = USE_AO_SPECIALBLUR && m_config.blur ? ...;
shader
  .bindViewNormalTexture(...)
  .bindControlBuffer(hbaoControlBuffer);

#if USE_AO_LAYERED_SINGLEPASS
shader.bindLinearDepthTexture(...);
#if USE_AO_LAYERED_SINGLEPASS == AO_LAYERED_IMAGE
shader.bindOutputImage(..., USE_AO_SPECIALBLUR ? 
  Mn::GL::ImageFormat::RG16F : Mn::GL::ImageFormat::R8)
#endif
shader.draw(fullScreenTriangles);
#if USE_AO_LAYERED_SINGLEPASS == AO_LAYERED_IMAGE
Mn::GL::Renderer::setMemoryBarrier(
  Mn::GL::Renderer::MemoryBarrier::TextureFetch|
  Mn::GL::Renderer::MemoryBarrier::ShaderImageAccess);
#endif
#else
for (int i = 0; i < HBAO_RANDOM_ELEMENTS; i++) {
  hbao2CalcFramebuffer.attachTextureLayer(textures.hbao2_resultarray, 0, i);

  shader
    .setFloat2Offset({float(i % 4) + 0.5f, float(i / 4) + 0.5f})
    .setJitter(m_hbaoRandom[i].get_value())
    .bindLinearDepthTexture(textures.hbao2_depthview[i])
    .draw(fullScreenTriangle);
}
#endif
```

and similarly in other parts. Various renderer state and enable/disable bits are all in `GL::Render`, again use the search functionality to look for matching APIs.